src/Symfony/Component/BrowserKit/Request.php

```diff
@@ -39,7 +39,7 @@ public function __construct(string $uri, string $method, array $parameters = [],
         $this->method = $method;
 
         array_walk_recursive($parameters, static function (&$value) {
-            $value = (string) $value;
+            $value = \is_bool($value) ? $value : (string) $value;
```

Comment:
this is a no-go. A query string cannot contain a boolean value. The string conversion was totally intended

Reply:
Yeah, you are right. 

I have edited code. We should cast `false`-boolean to `'0'`-string, because `symfony/validator` does not handle the empty string as `false`:
https://github.com/symfony/symfony/blob/19940ffcfe9bbae825ca951f1bb4c78a8a84f9ad/src/Symfony/Component/Validator/Constraints/IsFalseValidator.php#L29

Reply:
How does that change what @stof wrote? If you want the request to contain `'0'`, use `'0'`.

Reply:
> How does that change what @stof wrote? If you want the request to contain `'0'`, use `'0'`.

I want use the boolean `false`-value, but casting to the empty string breaks logic. I try send the `false`, but request cast to the empty string and validator fails.

Reply:
> I want use the boolean `false`-value

Okay, but as @stof wrote:

> The query string (and same for the form-url-encoded body) cannot contain a boolean value.

So why would you want to use a value in a request that a real HTTP request could never transport?

Reply:
> So why would you want to use a value in a request that a real HTTP request could never transport?

I use the test framework Codeception with a module [`codeception/module-rest`](https://codeception.com/docs/APITesting) for API tests. 

When I try send request with the `false`-value, the Request instance casts the value to empty string and validation fails, because the empty string is not a false-value in the Symfony validator and not a boolean value in the Laravel.

I have followed the stack and saw that the `Symfony\Component\BrowserKit\Request` casts all parameters to a string. There is a reason why I made the PR. I understand we should follow  the specification, but the PHP-casting `false` to the empty string is not so correct for validators - [Symfony/IsFalseValidator](https://github.com/symfony/symfony/blob/19940ffcfe9bbae825ca951f1bb4c78a8a84f9ad/src/Symfony/Component/Validator/Constraints/IsFalseValidator.php#L29) and [Laravel/IsBoolean](https://github.com/illuminate/validation/blob/aff6df3c87bd942c3d37373291aa1c527bc8fa65/Concerns/ValidatesAttributes.php#L438).

Reply:
If `codeception/module-rest` expects the BrowserKit Request parameters to contain the non-encoded JSON data, it means that it changes the meaning of that field compared to what it means in Symfony, which is clearly a bug in that package. That's not how json requests are represented in `AbstractBrowser::jsonRequest`.

If `codeception/module-rest` misuses the component, you should open an issue (or a PR) for them, not breaking the normal usage of the Symfony component.

Reply:
> If codeception/module-rest misuses the component, you should open an issue (or a PR) for them, not breaking the normal usage of the Symfony component.

I do it, if the PR will rejected. I opened the PR in the Symfony repo, because the casting happens inside `Symfony\Component\BrowserKit\Request`.

Reply:
But as said above, the casting is totally legit.

Reply:
> But as said above, the casting is totally legit.

Yeah. I agree. The main problem is PHP behaviour when casts `false` to a string. 

Thank you for patience and answers!